File: BST_iterator.cpp
// Problem Link - https://www.interviewbit.com/problems/bst-iterator/submissions/
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
stack<TreeNode*> A;
void inordertraversal(TreeNode* head){
    while(head != NULL)
    {
        A.push(head);
        head = head->left;
    }
    return;
}

BSTIterator::BSTIterator(TreeNode *root) {
    inordertraversal(root);
}

/** @return whether we have a next smallest number */
bool BSTIterator::hasNext() {
    return !A.empty();
}

/** @return the next smallest number */
int BSTIterator::next() {
    TreeNode* cr = A.top(); A.pop();
    if(cr->right != NULL) inordertraversal(cr->right);
    return cr->val;
}

/**
 * Your BSTIterator will be called like this:
 * BSTIterator i = BSTIterator(root);
 * while (i.hasNext()) cout << i.next();
**/

File: striver_sheet/binary_search_tree_part_II/bst_iterator.py
# Leetcode solution
# Link : https://leetcode.com/problems/binary-search-tree-iterator/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:
    def __init__(self, root: Optional[TreeNode]):
        self.st = []
        self.pushAll(root)

    def next(self) -> int:
        x = self.st.pop()
        self.pushAll(x.right)
        return x.val

    def hasNext(self) -> bool:
        x = len(self.st)
        if x > 0:
            return True
        return False

    def pushAll(self, root) -> None:
        while root != None:
            self.st.append(root)
            root = root.left


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()
